[Bookings] Clear filters action

irfano opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses, e.g., WOOMOB-373. -->
Closes WOOMOB-1473

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This adds a CLEAR button to the Bookings filter screen and updates the filter logic and data model for the clear function.

I changed my mind and added more commits after my original plan, so I suggest not following the commits.
For my last commit (23c9069722718f23e13a8175826cbadb94058f20), I removed `BookingType.ANY` and use `null` for the same purpose. I believe this approach is cleaner, though I’m not strongly opinionated about it. Feel free to share your thoughts.

The style of the CLEAR button should match the one used on the Order filters screen.

### Test Steps
<!-- Describe how to test your changes. Include only what’s needed for the reviewer to validate the behavior.
For new features, outline the main user flow or goal rather than every tap or click — the reviewer should be able to complete the flow naturally.
If applicable, mention key devices, scenarios, or edge cases to verify. -->
1. Log in with a CIAB store.
2. Go to Bookings.
3. Go to the All tab.
4. Tap the Filters button.
5. Select Booking type.
6. Select a type.
7. Navigate back.
8. Tap the CLEAR button.
9. Tap the X button to verify that the unsaved changes dialog is displayed.
10. Tap KEEP CHANGES.
11. Tap Show bookings button.
12. Verify that the filter has been reset.

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->
<img width="360" src="https://github.com/user-attachments/assets/4361a6cf-6c21-4cb3-b9b9-10d3ba6c678d" />

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

irfano committed:
Add clear button string for booking filters

Adds a new string resource `bookings_filters_clear_button` for the "CLEAR" button in the bookings filters UI.

irfano committed:
Add clear button to booking filters screen

This commit adds a "Clear" button to the booking filters screen.

irfano committed:
Use "Any" as the default for booking type filter

This change ensures that the booking type filter defaults to "Any" instead of being nullable.

irfano committed:
Simplify booking filter state management

This commit refactors the state management for the booking filters screen.

irfano committed:
Reset booking filters on clear

When clearing the booking filters, reset the `updatedBookingFilters` in the UI state to a new empty `BookingFilters` instance.

irfano committed:
Simplify `BookingType` import in `BookingFilterListViewModelTest`

irfano committed:
Test onClear on the booking filter list screen

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/bookings/filter/BookingFilterListUiState.kt:
```diff
@@ -21,15 +21,22 @@ enum class BookingFilterPage {
 
+/**
+ * UI state for the Bookings Filters screen.
+ *
+ * @property initialBookingFilters The stored filters loaded from storage DataStore and shown when the screen opens.
+ * @property updatedBookingFilters The in-memory working copy reflecting user changes; not persisted until the user saves.
+ */
 data class BookingFilterListUiState(
-    val initialBookingFilters: BookingFilters? = null,
-    val newBookingFilters: Set<BookingsFilterOption> = emptySet(),
+    val initialBookingFilters: BookingFilters = BookingFilters(),
+    val updatedBookingFilters: BookingFilters = initialBookingFilters,
```
I changed `newBookingFilters` to `updatedBookingFilters`. Previously, `newBookingFilters` stored only the modified in-memory filter options. Now, `updatedBookingFilters` holds all options. This change simplifies clearing filters.

irfano replied:
There’s another weakness in the previous approach (`newBookingFilters: Set<BookingsFilterOption>`). Suppose the filter already contains two customers and I remove one. The previous approach can’t represent that change, whereas `updatedBookingFilters` always reflects the latest filter state.

irfano committed:
Fix test in BookingListViewModelTest

Replaced `eq()` with `argThat()` in the `BookingListViewModelTest` to make the verification of filters less strict.

irfano committed:
Make Booking Type filter nullable again

Previously, the booking type filter always had a default value (`ANY`) and could not be removed from the saved filters. This commit updates the filter to be nullable, aligning its behavior with other booking filters.

irfano committed:
Fix: Use updated booking filters on booking type filter route

In the `BookingFilterListScreen`, the `BookingTypeFilterRoute` was incorrectly using the `initialBookingFilters` to set its initial type. This has been corrected to use `updatedBookingFilters` to ensure the filter reflects the current selections.

irfano committed:
Refactor booking type filter to use null for "Any"

Instead of using a dedicated `BookingType.Type.ANY` enum value to represent the "Any" filter option, the `BookingType.value` is now nullable. A `null` value now signifies that no specific booking type filter is applied.

This change simplifies the logic by removing the `ANY` type and aligning the implementation with the concept that a `null` filter value means no filtering.

The changes are reflected across:
- `BookingsFilterOption`: `BookingType` now accepts a nullable `Type`, and the `ANY` enum is removed.
- `BookingFilterRepository`: Logic updated to handle nullable `bookingType.value`.
- `BookingTypeFilterUiState`: Default selection is now `BookingType(null)` and `titleRes` handles the null case.
- `BookingTypeFilterViewModel`: Initial state is updated to use `BookingType(null)`.

JorgeMucientes commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/bookings/BookingsFilterOption.kt:
```diff
@@ -9,8 +17,13 @@ sealed interface BookingsFilterOption {
+    /**
+     * Booking type filter.
+     *
+     * [value] == null means “Any” (no filter is applied); otherwise a concrete [Type] is selected.
```
Change sounds good to me

JorgeMucientes commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/bookings/filter/BookingFilterListUiState.kt:
```diff
@@ -21,15 +21,23 @@ enum class BookingFilterPage {
     Location,
 }
 
+/**
+ * UI state for the Bookings Filters screen.
+ *
+ * @property initialBookingFilters The stored filters loaded from storage DataStore and shown when the screen opens.
+ * @property updatedBookingFilters The in-memory working copy reflecting user changes; not persisted until the user
+ * saves.
+ */
```
Do we need this comments? We usually don't add this level of commenting in the code unless some extra context is required for something not obvious. 
Overall I feel with the use of AI we might be tempted to add more comments to the code in areas where things are fairly clear. Idk if this particular case was AI generated but just saying as something to keep in mind.

irfano replied:
I used a specific AI prompt to generate these comments for just the two parameters that I thought were necessary. Basically, `initialBookingFilters` comes from the store, while `updatedBookingFilters` represents the temporary in-memory filters. Since I changed the behavior from  `newBookingFilters` to `updatedBookingFilters`. I think these comments could be useful, but if you feel the code is already clear enough, I can remove them.

irfano replied:
I’ve merged this, but I can remove the comments in my following PRs if you think we should.

JorgeMucientes commented on WooCommerce/src/main/res/values/strings.xml:
```diff
@@ -4250,6 +4250,7 @@
     <string name="bookings_filter_type_service">Service</string>
     <string name="bookings_filter_type_event">Event</string>
     <string name="bookings_filters_selected_option_content_description">Selected filter option</string>
+    <string name="bookings_filters_clear_button">CLEAR</string>
```
We may want to use existing token `<string name="clear">Clear</string>` to avoid adding an unnecessary translation? And then add `.uppercase()` call if we want to display with upper case style

irfano replied:
I avoided using `uppercase()` in the UI layer. Plus, using a common string across different screens can cause issues, in our case we don't use `R.string.clear` in many places. Still, I wasn't sure about my decision. I applied your suggestion: 8163a9e1df02c9376f547e2490a8f39d5c84d921

JorgeMucientes reviewed (approved):
Nicely done @irfano. Just left a couple minor np in case you want to address them.

irfano committed:
Use common "clear" string resource in booking filters

Replaces the specific `bookings_filters_clear_button` string with the general `clear` string in the booking filter list screen.

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-android/milestone/275">23.7</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

irfano committed:
Correct import order in BookingFilterListScreen
